test/compile/test_compilation.py

```diff
@@ -0,0 +1,183 @@
+    def test_decompile_success_cached(
+        self: CompilationManagerSpec,
+        mock_path_exists: MagicMock,
+    ) -> None:
+        mock_path_exists.return_value = True
+
+        folder = self.compilation_manager.decompile(self.swf)
+
+        assert mock_path_exists.call_count == 3
+        assert folder == Path('.Patcher-Temp/ORSXG5BOON3WM===')
+        self.mock_decompiler.assert_not_called()
+
+    @patch('pathlib.Path.exists')
+    def test_decompile_failure_no_input(
+        self: CompilationManagerSpec,
+        mock_path_exists: MagicMock,
+    ) -> None:
+        mock_path_exists.side_effect = [True, False]
+
+        with raises(FileNotFoundError):
+            self.compilation_manager.decompile(self.swf)
+
+        assert mock_path_exists.call_count == 2
+        self.mock_decompiler.assert_not_called()
+
+    @patch('pathlib.Path.exists')
+    def test_decompile_failure_jpexs_error(
+        self: CompilationManagerSpec,
+        mock_path_exists: MagicMock,
+    ) -> None:
+        mock_path_exists.return_value = True
+        self.mock_decompiler.export_scripts.return_value = False
+
+        with raises(DependencyError):
+            self.compilation_manager.decompile(self.swf, drop_cache=True)
+
+        assert mock_path_exists.call_count == 3
+        self.mock_decompiler.export_scripts.assert_called_once_with(
+            self.swf, Path('.Patcher-Temp/ORSXG5BOON3WM===')
+        )
+
+    def test_recompile_with_check_success(
+        self: CompilationManagerSpec,
+    ) -> None:
+        self.mock_decompiler.recompile_data.return_value = True
+
+        self.compilation_manager.recompile_with_check("Script", self.folder, self.swf, self.swf)
+
+        self.mock_decompiler.recompile_data.assert_called_once_with(
+            "Script", self.folder, self.swf, self.swf
+        )
+
+    def test_recompile_with_check_failure(
+        self: CompilationManagerSpec,
+    ) -> None:
+        self.mock_decompiler.recompile_data.return_value = False
+
+        with raises(DependencyError):
+            self.compilation_manager.recompile_with_check(
+                "Script", self.folder, self.swf, self.swf
+            )
+
+        self.mock_decompiler.recompile_data.assert_called_once_with(
+            "Script", self.folder, self.swf, self.swf
+        )
+
+    def test_recompile_success_xml_mode(
+            self: CompilationManagerSpec,
+        ) -> None:
+        self.compilation_manager.recompile(self.folder, self.folder, self.swf, xml_mode=True)
+        self.mock_decompiler.rebuild_xml.assert_called_once_with(self.folder, self.swf)
+
+    @patch('compile.compilation.CompilationManager.recompile_with_check')
+    def test_recompile_success_full(
+        self: CompilationManagerSpec,
+        mock_recompile_with_check: MagicMock
+    ) -> None:
+        self.compilation_manager.recompile(
+            self.folder, self.swf, self.swf, recompile_all=True
+        )
+
+        expected_calls = [
```

Comment:
Maybe a for loop could be used here?

Reply:
Sounds good!

Reply:
edit: using `assert_has_calls` allows us to verify that the calls were made in the specified order, so it seems to be the better function for this use case.